ygen/gogen.go

```diff
@@ -930,13 +981,81 @@ func writeGoStruct(targetStruct *yangDirectory, goStructElements map[string]*yan
 //   }
 func generateValidator(buf *bytes.Buffer, structDef generatedGoStruct) error {
 	if err := goTemplates["structValidator"].Execute(buf, structDef); err != nil {
-		fmt.Println(err)
 		return err
 	}
 
 	return nil
 }
 
+// generateGetListKey generates a function extracting the keys from a list
+// defined in the yangDirectory s, and appends it to the supplier buffer. The
+// nameMap stores maps between the key YANG field identifiers and their Go
+// identifiers.
+//
+// If the input yangDirectory is the following list entry:
+//
+//  list foo {
+//    key "bar baz";
+//
+//    leaf bar { type string; }
+//    leaf baz { type uint8; }
+//    leaf colour { type string; }
+//  }
+//
+// Which is mapped into the Go struct:
+//
+//  type Foo {
+//    Bar *string `path:"bar"`
+//    Baz *uint8  `path:"baz"`
+//    Colour *string `path:"colour"`
+//  }
+//
+// The generated method will;
+//  - Check pointer keys to ensure they are non-nil.
+//  - Return a map[string]interface{} keyed by the name of the key in the YANG schema, with the value
+//    specified in the struct.
+//
+// i.e.: for the above struct:
+//
+//  func (t *Foo) ΛListKeyMap() (map[string]interface{}, error) {
+//	if t.Bar == nil {
+//	   return nil, fmt.Errorf("key value for Bar is nil")
+//	}
+//
+//	if t.Baz == nil {
+//	   return nil, fmt.Errorf("key value for Baz is nil")
+//	}
+//
+//	return map[string]interface{}{
+//	  "bar": *t.Bar,
+//	  "baz": *t.Baz,
+//	}
+//  }
+func generateGetListKey(buf *bytes.Buffer, s *yangDirectory, nameMap map[string]*yangFieldMap) error {
+	if s.listAttr == nil {
+		return nil
```

Comment:
Is this an error?

Reply:
It's not -- I'll add a comment. This was taking a pattern that you've suggested before, whereby we perform the checks of whether the function is relevant inside the function. In this case, if the `yangDirectory` was an unkeyed list, or just a container, then we just return with nothing to do.